### PR TITLE
[WIP] Issues/63 support maven toolchains plugin

### PR DESCRIPTION
**Pull request by exabrial**

### CODE REVIEW

**Comment by exabrial** [comment]
How would one assert which jdk is used, especially since antbuilder.java is a static method? (normally i'd mock out that dependency and check to see if the jvm flag is passed)

**Comment by hazendaz** [comment]
@exabrial Look at the IT tests and see if you can create one there that runs with toolchains.  I did go ahead and release 3.1.5 so others can use that but need to hold on this until concerns raised are addressed :)

**Comment by exabrial** [comment]
Attempting an IT... I'm curious what's going to happen on CI

File: src/main/groovy/org/codehaus/mojo/spotbugs/SpotBugsMojo.groovy [hunk 1]
```diff
@@ -998,31 +1014,43 @@ class SpotBugsMojo extends AbstractMavenReport implements SpotBugsPluginsTrait {
 
         def ant = new AntBuilder()
 
-        log.info("Fork Value is ${fork}")
```
Comment (KengoTODA) [inline src/main/groovy/org/codehaus/mojo/spotbugs/SpotBugsMojo.groovy]: please do not remove log that is not related with your fix.

File: src/main/groovy/org/codehaus/mojo/spotbugs/SpotBugsMojo.groovy [hunk 2]
```diff
@@ -998,31 +1014,43 @@ class SpotBugsMojo extends AbstractMavenReport implements SpotBugsPluginsTrait {
...
+        if (sourceEncoding) {
+            effectiveEncoding = sourceEncoding
+        }
+            
+        def jvmExecutable = null;
+        if (toolchainManager != null && fork) {
+            def tc = toolchainManager.getToolchainFromBuildContext("jdk", session);
+              if (tc != null) {
+                String foundExecutable = tc.findTool("java");
+                if (foundExecutable != null) {
```
Comment (KengoTODA) [inline src/main/groovy/org/codehaus/mojo/spotbugs/SpotBugsMojo.groovy]: We can shorten implementation by [safe navigation operator](https://en.wikipedia.org/wiki/Safe_navigation_operator).

File: src/main/groovy/org/codehaus/mojo/spotbugs/SpotBugsMojo.groovy (hunk 1 above)
Comment (KengoTODA) [reply src/main/groovy/org/codehaus/mojo/spotbugs/SpotBugsMojo.groovy]: I found that this log is moved to L1041, what is our motivation?

File: src/main/groovy/org/codehaus/mojo/spotbugs/SpotBugsMojo.groovy [hunk 3]
```diff
@@ -998,31 +1014,43 @@ class SpotBugsMojo extends AbstractMavenReport implements SpotBugsPluginsTrait {
...
+                }
+              }
+        }
+            
+        log.info("Fork Value is ${fork}")
 
-        ant.java(classname: "edu.umd.cs.findbugs.FindBugs2", inputstring: getSpotbugsAuxClasspath(), fork: "${fork}", failonerror: "true", clonevm: "false", timeout: "${timeout}", maxmemory: "${maxHeap}m") {
+        ant.java([classname: "edu.umd.cs.findbugs.FindBugs2", inputstring: getSpotbugsAuxClasspath(), fork: "${fork}",
+            	jvm: jvmExecutable, failonerror: "true", clonevm: "false", timeout: "${timeout}", maxmemory: "${maxHeap}m",
+            ].findAll{it.value != null}) {
```
Comment (KengoTODA) [inline src/main/groovy/org/codehaus/mojo/spotbugs/SpotBugsMojo.groovy]: why we need this `findAll`?

File: src/main/groovy/org/codehaus/mojo/spotbugs/SpotBugsMojo.groovy (hunk 1 above)
Comment (exabrial) [reply src/main/groovy/org/codehaus/mojo/spotbugs/SpotBugsMojo.groovy]: No reason, just made a bit more sense in the log flow I thought

File: src/main/groovy/org/codehaus/mojo/spotbugs/SpotBugsMojo.groovy (hunk 2 above)
Comment (exabrial) [reply src/main/groovy/org/codehaus/mojo/spotbugs/SpotBugsMojo.groovy]: thanks. I don't know groovy

File: src/main/groovy/org/codehaus/mojo/spotbugs/SpotBugsMojo.groovy (hunk 3 above)
Comment (exabrial) [reply src/main/groovy/org/codehaus/mojo/spotbugs/SpotBugsMojo.groovy]: Because the `jvmExecutable` could possibly be null at this point. This ensures null values are dropped from the map. If you pass null to `ant.java` you get an NPE.

**Comment by exabrial** [comment]
btw I'm in another city at the moment, sorry for the slow progress. I don't have much time to work on this but I'll eventually get it working

**Comment by exabrial** [comment]
@KengoTODA Hey could this build possibly be failing because the lack of a toolchains.xml on the build machine? How can we add this file?

File: src/main/groovy/org/codehaus/mojo/spotbugs/SpotBugsMojo.groovy (hunk 2 above)
Comment (exabrial) [reply src/main/groovy/org/codehaus/mojo/spotbugs/SpotBugsMojo.groovy]: So actually, I need to leave the code this way to be consistent with how the other plugins generate their log messages about using Toolchains. I totally understand the safe operator though, and I'm glad you pointed that out.

**Comment by KengoTODA** [comment]
How about using `-s` option? It lets specify the path of `settings.xml`. Then you can commit `settings.xml` as a part of test case.

**Comment by exabrial** [comment]
@KengoTODA Just so you're absolutely informed what's going on, it's missing a `toolchains.xml` file, NOT the not `settings.xml` file you quoted above. The two files have different XSDs.

**Comment by exabrial** [comment]
The toolchains.xml is supposed to be a machine configuration of where the JVMs are physically located on disk. Including those in a commit means that everyone has to set up their machines in the exact same way. 
https://maven.apache.org/guides/mini/guide-using-toolchains.html

**Comment by KengoTODA** [comment]
How about --global-toolchains ?

**Comment by exabrial** [comment]
I'm going to give up :/ There's not a good way to do Integration Tests withtout control over the environment. I think the original problem I was facing is less of an issue anyway... (cross compiling a jdk 1.8 on a jdk 1.9 maven vm)

**Comment by hazendaz** [comment]
Tied to #63
